p2p/transport.go

```diff
@@ -52,6 +52,9 @@ type Transport interface {
 
 	// Dial connects to the Peer for the address.
 	Dial(NetAddress, peerConfig) (Peer, error)
+
+	// Cleanup any resources associated with Peer.
+	Cleanup(Peer)
```

Comment:
I think we should be extra careful when considering increasing coupling of our components. I like the fact that peer does not know anything about the transport.

Having Cleanup function do have cons:
- you can forget to call it

pros:
- very opaque
- and direct

Reply:
I agree with the desire for decoupling, but it seems in this case we have components that are genuinely coupled. The property that the peer doesn't know about the transport sounds nice in spirit, but in practice, as far as I understand here, the implementations of Peer and Transport are already tightly wound - a particular implementation of Transport returns a particular implementation of Peer, and a particular implementation of peer is expected to come from a particular implementation of Transport. So Peer in general doesn't need to know about Transport, but if the Transport that created it necessitates some cleanup, then it surely will.

By pushing this implicit coupling into an extra responsibility for callers, we're just asking them to forget about it. If we have transports that don't require this cleanup, we're cluttering the Transport interface unnecessarily. So I would think, in this case, `you can forget to call it` takes the cake.

We don't need to give the peer a direct reference to the transport. It could have some cleanup closure or a channel the Transport listens for.  But the current separation seems prone to misuse.

Case in point, it looks like we're forgetting to call Cleanup on peers in sw.OnStop

Reply:
That's fair. But the issue about making two calls still stands. Maybe we let Cleanup call Stop ?